sceptre/cli/diff.py

```diff
@@ -0,0 +1,85 @@
+import contextlib
+import json
+import sys
+from pathlib import Path
+from typing import Dict
+
+import click
+import yaml
+from click import Context
+
+from sceptre.cli.helpers import catch_exceptions
+from sceptre.context import SceptreContext
+from sceptre.diffing.diff_writer import DeepDiffWriter, DictDifferWriter, DiffLibWriter
+from sceptre.diffing.stack_differ import DeepDiffStackDiffer, DictDifferStackDiffer, DifflibStackDiffer, StackDiff
+from sceptre.plan.plan import SceptrePlan
+from sceptre.stack import Stack
+
+
+@contextlib.contextmanager
+def nullcontext(enter_result=None):
+    yield enter_result
+
+
+@click.command(name="diff", short_help="Compares deployed infrastructure with current configurations")
+@click.option(
+    '-o',
+    '--output-file',
+    type=click.Path(file_okay=True, dir_okay=False, resolve_path=True),
+    help=(
+        "If specified, the filepath that diff results will be printed to; "
+        "Otherwise they will streamed to stdout."
+    )
+)
+@click.option(
+    '-D',
+    '--differ',
+    type=click.Choice(['deepdiff', 'dictdiffer', 'difflib']),
+    default='deepdiff'
+)
+@click.argument('path')
+@click.pass_context
+@catch_exceptions
+def diff_command(ctx: Context, output_file: str, differ: str, path):
+    """This command is designed to indicate what the difference between the currently DEPLOYED
+    stack templates at the indicated command_path and the stack templates generated off of the
+    current templates and configurations.
+    """
+    context = SceptreContext(
+        command_path=path,
+        project_path=ctx.obj.get("project_path"),
+        user_variables=ctx.obj.get("user_variables"),
+        options=ctx.obj.get("options"),
+        ignore_dependencies=ctx.obj.get("ignore_dependencies"),
+        output_format=ctx.obj.get('output_format'),
+        no_colour=ctx.obj.get('no_colour')
+    )
+    output_format = context.output_format
+    plan = SceptrePlan(context)
+
+    if differ == "deepdiff":
+        stack_differ = DeepDiffStackDiffer()
+        writer_class = DeepDiffWriter
+    elif differ == 'dictdiffer':
+        stack_differ = DictDifferStackDiffer()
+        writer_class = DictDifferWriter
+    elif differ == 'difflib':
+        serializer = json.dumps if output_format == 'json' else yaml.dump
+        stack_differ = DifflibStackDiffer(serializer=serializer)
+        writer_class = DiffLibWriter
+    else:
+        raise ValueError(f"Unexpected differ type: {differ}")
+
+    diffs: Dict[Stack, StackDiff] = plan.diff(stack_differ)
+
+    if output_file:
+        path = Path(output_file)
+        path.parent.mkdir(parents=True, exist_ok=True)
+        context = path.open(mode='wt')
```

Comment:
So the concern that drove this is that other things that Sceptre does (especially with hooks) can put other things to stdout. If I just piped stdout to a file, it would also capture other junk I don't care about.

For example, I have a hook that packages SAM Templates. That hook outputs things to stdout. With all the things that could be configured in a Sceptre project (especially re: hooks), it seems like being able to dump a diff file would be really useful, so you can extract out JUST the output from this command and separate that from all the logs and other outputs that get spat out.

If you don't think I should do that, I suppose there are other ways, like updating all my hooks to redirect their output to stderr. But that would be a fair amount of work when we could avoid that with this simple option.

Reply:
You know, I actually looked into this and the only cases we have of hooks outputing to stdout only needed a single line change in one place (hurray for good design patterns). I've played around with this and I think I can pull out this option as you suggest. adding `> diff.txt` to the end of the sceptre command is sufficient to pipe it to a file. I'll stick with the Sceptre convention of stdout only.